crates/sparrow-scheduler/src/pipeline.rs

```diff
@@ -0,0 +1,109 @@
+use sparrow_arrow::Batch;
+
+use crate::queue::Queue;
+use crate::{Partition, TaskRef, Worker};
+
+/// A push-based interface used by the scheduler to drive query execution
+///
+/// A pipeline processes data from one or more input partitions, producing output
+/// to one or more output partitions. As a [`Pipeline`] may drawn on input from
+/// more than one upstream [`Pipeline`], input partitions are identified by both
+/// a child index, and a partition index, whereas output partitions are only
+/// identified by a partition index.
+///
+/// This is not intended as an eventual replacement for the physical plan
+/// representation, but rather a generic interface that parts of the physical
+/// plan are converted to for execution.
+///
+/// # Eager vs Lazy Execution
+///
+/// Whether computation is eagerly done on push, or lazily done on pull, is
+/// intentionally left as an implementation detail of the [`Pipeline`]
+///
+/// This allows flexibility to support the following different patterns, and potentially more:
+///
+/// An eager, push-based pipeline, that processes a batch synchronously in [`Pipeline::push`]
+/// and immediately wakes the corresponding output partition.
+///
+/// A parallel, push-based pipeline, that enqueues the processing of a batch to
+/// the thread pool in [`Pipeline::push`], and wakes the corresponding output
+/// partition when the job completes. Order and non-order preserving variants
+/// are possible
+///
+/// A merge pipeline which combines data from one or more input partitions into one or
+/// more output partitions. [`Pipeline::push`] adds data to an input buffer, and wakes
+/// any output partitions that may now be able to make progress. This may be none if
+/// the operator is waiting on data from a different input partition.
+///
+/// An aggregation pipeline which combines data from one or more input partitions into
+/// a single output partition. [`Pipeline::push`] would eagerly update the computed
+/// aggregates, and the final [`Pipeline::close`] trigger flushing these to the output.
+/// It would also be possible to flush once the partial aggregates reach a certain size.
+///
+/// A partition-aware aggregation pipeline, which functions similarly to the above, but
+/// computes aggregations per input partition, before combining these prior to flush.
+///
+/// An async input pipeline, which has no inputs, and wakes the output partition
+/// whenever new data is available.
+///
+/// A JIT compiled sequence of synchronous operators, that perform multiple operations
+/// from the physical plan as a single [`Pipeline`]. Parallelized implementations
+/// are also possible.
+///
+pub trait Pipeline: Send + Sync + std::fmt::Debug {
+    type Context: error_stack::Context;
```

Comment:
To discuss:

Should we have a single `sparrow-scheduler::Error` or `sparrow-scheduler::PipelineError` that we require *all* pipelines produce? Currently we don't, which creates some pain. Specifically, the `Pipeline` trait is not "object safe", meaning we need to wrap it into a *different* trait before we put it behind `dyn`. See some of the complexity / cruft in `task.rs`.

Reply:
I think if we like that, it is pretty easy to add and cleans up code significantly, so I'll consider this something to do in this PR.